kernel/src/grant.rs

```diff
@@ -352,34 +257,30 @@ pub struct Iter<'a, T: 'a + Default> {
         core::slice::Iter<'a, Option<&'static dyn ProcessType>>,
         fn(&Option<&'static dyn ProcessType>) -> Option<&'static dyn ProcessType>,
     >,
+    skips_allowed: bool,
 }
 
-impl<T: Default> Iterator for Iter<'_, T> {
-    type Item = AppliedGrant<T>;
+impl<'a, T: Default> Iterator for Iter<'a, T> {
+    type Item = AppliedGrant<'a, T>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        // Save a local copy of grant_num so we don't have to access `self`
-        // in the closure below.
-        let grant_num = self.grant.grant_num;
-
+        let grant = self.grant;
         // Get the next `AppId` from the kernel processes array that is setup to use this grant.
         // Since the iterator itself is saved calling this function
         // again will start where we left off.
-        let res = self.subiter.find(|process| {
-            // We have found a candidate process that exists in the
-            // processes array. Now we have to check if this grant is setup
-            // for this process. If not, we have to skip it and keep
-            // looking.
-            if let Some(grant_ptr) = process.get_grant_ptr(grant_num) {
-                !grant_ptr.is_null()
-            } else {
-                false
-            }
-        });
-
-        // Check if our find above returned another `AppId`, or if we hit the
-        // end of the iterator. If we found another app, try to access its grant
-        // region.
-        res.map_or(None, |process| self.grant.grant(process.appid()))
+        let skips_allowed = self.skips_allowed;
+        self.subiter.find_map(
+            |process| match AppliedGrant::get_if_allocated(grant, process) {
+                Ok(ag) => Some(ag),
+                Err(None) => None,
+                Err(Some(_err)) => {
+                    if skips_allowed {
+                        None
+                    } else {
+                        panic!("Attempted to re-enter a grant region.")
```

Comment:
```suggestion
                        // Panic if the capsule asked us to enter an already entered
                        // grant region. This preserves type correctness (but does crash
                        // the system).
                        //
                        // This panic represents a tradeoff. While it is undesirable to have the
                        // potential for a runtime crash in this grant region code, it balances
                        // usability with type correctness. The challenge is that calling `self.apps.iter()`
                        // is a common pattern in capsules to access the grant region of every
                        // app that is using the capsule, and sometimes it is intuitive to call that
                        // inside of a `self.apps.enter(app_id, |app| {...})` closure. However, `.enter()`
                        // means that app's grant region is entered, and then a naive `.iter()` would
                        // re-enter the grant region and cause undefined behavior. We considered
                        // different options to resolve this.
                        //
                        // 1. Have `.iter()` only iterate over grant regions which are not entered.
                        //    This avoids the bug, but could lead to unexpected behavior, as `self.apps.iter()` will
                        //    do different things depending on where in a capsule it is called.
                        // 2. Have the compiler detect when `.iter()` is called when a grant region has
                        //    already been entered. We don't know of a viable way to implement this.
                        // 3. Panic if `.iter()` is called when a grant is already entered.
                        //
                        // We decided on option 3 because it balances minimizing surprises (`self.apps.iter()`
                        // will always iterate all grants) while also protecting against the bug. We expect
                        // that any code that attempts to call `self.apps.iter()` after calling `.enter()` will
                        // immediately encounter this `panic!()` and have to be refactored before any tests
                        // will be successful. Therefore, this `panic!()` should only occur at development/testing
                        // time.
                        //
                        // ## How to fix this error
                        //
                        // If you are seeing this panic, you need to refactor your capsule to not
                        // call `.iter()` or `.each()` from inside a `.enter()` closure. That is, you
                        // need to close the grant region you are currently in before trying to iterate
                        // over all grant regions.
                        panic!("Attempted to re-enter a grant region.")
```